### ISSUE TITLE
segfaults after MDB_MAP_FULL

### ISSUE THREAD

**Issue by phraktle** [lmdbjava/lmdbjava#17]
The following code segfaults, after the db is full.

```java

import static com.jakewharton.byteunits.BinaryByteUnit.MEBIBYTES;

import java.io.File;
import java.nio.ByteBuffer;
import java.util.concurrent.ThreadLocalRandom;

import org.lmdbjava.Dbi;
import org.lmdbjava.DbiFlags;
import org.lmdbjava.Env;

public class LmdbFullTest {

    public static void main(String[] args) {
        File path = new File("/tmp/lmdb_full/");
        path.mkdir();
        for (;;) {
            writeFull(path);
        }
    }

    private static void writeFull(File path) {
        Env<ByteBuffer> env = Env.create().setMapSize(MEBIBYTES.toBytes(10)).setMaxDbs(1).open(path);
        Dbi<ByteBuffer> db = env.openDbi("test", DbiFlags.MDB_CREATE);

        try {
            byte[] k = new byte[64];
            ByteBuffer key = ByteBuffer.allocateDirect(64);
            ByteBuffer val = ByteBuffer.allocateDirect(1024);

            ThreadLocalRandom rnd = ThreadLocalRandom.current();
            int count = 0;
            for (;;) {
                rnd.nextBytes(k);
                key.clear();
                key.put(k).flip();
                val.clear();
                db.put(key, val);
                System.out.println("written " + ++count);
            }
        } catch (Exception e) {
            e.printStackTrace(System.out);
        }

        System.out.println("closing db");
        db.close();

        System.out.println("closing env");
        env.close();
    }

}
```

**Comment by kk00ss** [comment]
No segfault happens on my machine with your code, I'm using windows however. What OS do you use ?

**Comment by phraktle** [comment]
Hi @kk00ss, I have tested this on Mac OS X, but we have seen segfaults on/after `MDB_MAP_FULL` on Linux as well.

**Comment by kk00ss** [comment]
There is another java driver for LMDB - https://github.com/deephacks/lmdbjni be aware it uses Unsafe. It doesn't fail with segfault on my machine after MDB_MAP_FULL, but there is different problem with it (it stops recognizing mdb file as correct lmdb database)

**Comment by benalexau** [comment]
```

package au.com.acegi.core.fff;

import static com.jakewharton.byteunits.BinaryByteUnit.MEBIBYTES;
import java.io.File;
import java.nio.ByteBuffer;
import java.util.concurrent.ThreadLocalRandom;
import org.junit.Test;
import org.lmdbjava.Dbi;
import org.lmdbjava.DbiFlags;
import org.lmdbjava.Env;
import org.lmdbjava.EnvFlags;

public final class Issue17 {

  @Test
  public void test() {
    final File path = new File("lmdbTest");
    path.deleteOnExit();

    Env<ByteBuffer> env = Env.create().setMapSize(MEBIBYTES.toBytes(10))
        .setMaxDbs(1).open(path, EnvFlags.MDB_NOSUBDIR);
    Dbi<ByteBuffer> db = env.openDbi("test", DbiFlags.MDB_CREATE);

    try {
      byte[] k = new byte[64];
      ByteBuffer key = ByteBuffer.allocateDirect(64);
      ByteBuffer val = ByteBuffer.allocateDirect(1024);

      ThreadLocalRandom rnd = ThreadLocalRandom.current();
      int count = 0;
      for (;;) {
        rnd.nextBytes(k);
        key.clear();
        key.put(k).flip();
        val.clear();
        db.put(key, val);
        System.out.println("written " + ++count);
      }
    } catch (Exception e) {
      e.printStackTrace(System.out);
    }

    System.out.println("closing db");
    db.close();

    System.out.println("closing env");
    env.close();
  }

}
```

Gives on my Linux machine with released LmdbJava 0.0.2:

```
.....
written 4898
written 4899
org.lmdbjava.Env$MapFullException: Environment mapsize reached (-30792)
	at org.lmdbjava.ResultCodeMapper.<clinit>(ResultCodeMapper.java:63)
	at org.lmdbjava.Env$Builder.open(Env.java:369)
	at org.lmdbjava.Env$Builder.open(Env.java:388)
	at au.com.acegi.core.fff.Issue17.test(Issue17.java:22)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:367)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:274)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
closing db
closing env
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 32.719 sec - in au.com.acegi.core.fff.Issue17
```

To move this forward, would it be possible to produce a minimal test (eg one that uses larger keys so there's fewer iterations required) which still seg faults on your machine and submit it as a PR (or just to this ticket) and I'll add it to `EnvTest` or similar to verify against the various CI environments.

**Comment by benalexau** [comment]
Moving further discussion to PR #19 as it offers us some automated building convenience via the CI environment.
